saq/job.py

```diff
@@ -258,3 +268,77 @@ def replace(self, job):
         """Replace current attributes with job attributes."""
         for field in job.__dataclass_fields__:
             setattr(self, field, getattr(job, field))
+
+
+class PeriodicHeartbeat:
+    """Class to manage agent Heartbeats while Job is running"""
+
+    def __init__(self, job: Job) -> None:
+        self._running: bool = True
+        self._heartbeat_task: typing.Optional[asyncio.Task] = None
+        self.job = job
+        self.heartbeat_enabled: bool = job.heartbeat and job.heartbeat > 0
+        self.time_between_ticks = (
+            max(round(job.heartbeat / 2), 1) if self.heartbeat_enabled > 0 else 0
+        )
+
+    async def start(self, func: typing.Awaitable):
+        """start
+
+        Executes function with a periodic heartbeat.
+        Args:
+            func (typing.Awaitable): _description_
+
+        Returns:
+            _type_: _description_
+        """
+        if self.heartbeat_enabled:
+            logger.info(
+                "starting heartbeat service for %s. Ticking every %s second(s)",
+                self.job.id,
+                self.time_between_ticks,
+            )
+        self._heartbeat_task = asyncio.create_task(self._periodically_publish())
+        executed_func: typing.Optional[typing.Any] = None
+        try:
+            executed_func = await func
+            self._heartbeat_task.cancel()
+            await self._heartbeat_task
+        except asyncio.CancelledError:
+            pass
+        return executed_func
+
+    async def stop(self):
+        """Stop heartbeat service."""
+        if self.heartbeat_enabled:
+            logger.info("stopping heartbeat service for %s", self.job.id)
+        if self._running:
+            self._running = False
+
+    async def _periodically_publish(self):
+        """Periodically publish heartbeat"""
+        while self._running and self.heartbeat_enabled:
+            logger.info(
+                "ticking heartbeat for %s, and sleeping for %s second(s)",
+                self.job.id,
+                self.time_between_ticks,
+            )
+            await self.job.update()
+            await asyncio.sleep(self.time_between_ticks)
+
+
+# decorator to perform a heartbeat in the background while a function is running
+def monitored_job(func):
```

Comment:
"monitored_job" feels like a misnomer.

This doesn't mean that the job is now monitored... all jobs are "monitored". 

Is there a better name here?

Reply:
would something like `periodic_heartbeat` be more appropriate?